File: distribution/python/dqops/client/models/trino_parameters_spec.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..models.aws_authentication_mode import AwsAuthenticationMode
from ..models.trino_engine_type import TrinoEngineType
from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.trino_parameters_spec_properties import TrinoParametersSpecProperties


T = TypeVar("T", bound="TrinoParametersSpec")


@_attrs_define
class TrinoParametersSpec:
    """
    Attributes:
        trino_engine_type (Union[Unset, TrinoEngineType]):
        host (Union[Unset, str]): Trino host name. Supports also a ${TRINO_HOST} configuration with a custom environment
            variable.
        port (Union[Unset, str]): Trino port number. The default port is 8080. Supports also a ${TRINO_PORT}
            configuration with a custom environment variable.
        user (Union[Unset, str]): Trino user name. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use
            dynamic substitution.
        password (Union[Unset, str]): Trino database password. The value can be in the ${ENVIRONMENT_VARIABLE_NAME}
            format to use dynamic substitution.
        aws_authentication_mode (Union[Unset, AwsAuthenticationMode]):
        athena_region (Union[Unset, str]): The AWS Region where queries will be run. Supports also a ${ATHENA_REGION}
            configuration with a custom environment variable.
        catalog (Union[Unset, str]): The catalog that contains the databases and the tables that will be accessed with
            the driver. Supports also a ${TRINO_CATALOG} configuration with a custom environment variable.
        athena_work_group (Union[Unset, str]): The workgroup in which queries will run. Supports also a
            ${ATHENA_WORK_GROUP} configuration with a custom environment variable.
        athena_output_location (Union[Unset, str]): The location in Amazon S3 where query results will be stored.
            Supports also a ${ATHENA_OUTPUT_LOCATION} configuration with a custom environment variable.
        properties (Union[Unset, TrinoParametersSpecProperties]): A dictionary of custom JDBC parameters that are added
            to the JDBC connection string, a key/value dictionary.
        database (Union[Unset, str]):
    """

    trino_engine_type: Union[Unset, TrinoEngineType] = UNSET
    host: Union[Unset, str] = UNSET
    port: Union[Unset, str] = UNSET
    user: Union[Unset, str] = UNSET
    password: Union[Unset, str] = UNSET
    aws_authentication_mode: Union[Unset, AwsAuthenticationMode] = UNSET
    athena_region: Union[Unset, str] = UNSET
    catalog: Union[Unset, str] = UNSET
    athena_work_group: Union[Unset, str] = UNSET
    athena_output_location: Union[Unset, str] = UNSET
    properties: Union[Unset, "TrinoParametersSpecProperties"] = UNSET
    database: Union[Unset, str] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        trino_engine_type: Union[Unset, str] = UNSET
        if not isinstance(self.trino_engine_type, Unset):
            trino_engine_type = self.trino_engine_type.value

        host = self.host
        port = self.port
        user = self.user
        password = self.password
        aws_authentication_mode: Union[Unset, str] = UNSET
        if not isinstance(self.aws_authentication_mode, Unset):
            aws_authentication_mode = self.aws_authentication_mode.value

        athena_region = self.athena_region
        catalog = self.catalog
        athena_work_group = self.athena_work_group
        athena_output_location = self.athena_output_location
        properties: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.properties, Unset):
            properties = self.properties.to_dict()

        database = self.database

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if trino_engine_type is not UNSET:
            field_dict["trino_engine_type"] = trino_engine_type
        if host is not UNSET:
            field_dict["host"] = host
        if port is not UNSET:
            field_dict["port"] = port
        if user is not UNSET:
            field_dict["user"] = user
        if password is not UNSET:
            field_dict["password"] = password
        if aws_authentication_mode is not UNSET:
            field_dict["aws_authentication_mode"] = aws_authentication_mode
        if athena_region is not UNSET:
            field_dict["athena_region"] = athena_region
        if catalog is not UNSET:
            field_dict["catalog"] = catalog
        if athena_work_group is not UNSET:
            field_dict["athena_work_group"] = athena_work_group
        if athena_output_location is not UNSET:
            field_dict["athena_output_location"] = athena_output_location
        if properties is not UNSET:
            field_dict["properties"] = properties
        if database is not UNSET:
            field_dict["database"] = database

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.trino_parameters_spec_properties import (
            TrinoParametersSpecProperties,
        )

        d = src_dict.copy()
        _trino_engine_type = d.pop("trino_engine_type", UNSET)
        trino_engine_type: Union[Unset, TrinoEngineType]
        if isinstance(_trino_engine_type, Unset):
            trino_engine_type = UNSET
        else:
            trino_engine_type = TrinoEngineType(_trino_engine_type)

        host = d.pop("host", UNSET)

        port = d.pop("port", UNSET)

        user = d.pop("user", UNSET)

        password = d.pop("password", UNSET)

        _aws_authentication_mode = d.pop("aws_authentication_mode", UNSET)
        aws_authentication_mode: Union[Unset, AwsAuthenticationMode]
        if isinstance(_aws_authentication_mode, Unset):
            aws_authentication_mode = UNSET
        else:
            aws_authentication_mode = AwsAuthenticationMode(_aws_authentication_mode)

        athena_region = d.pop("athena_region", UNSET)

        catalog = d.pop("catalog", UNSET)

        athena_work_group = d.pop("athena_work_group", UNSET)

        athena_output_location = d.pop("athena_output_location", UNSET)

        _properties = d.pop("properties", UNSET)
        properties: Union[Unset, TrinoParametersSpecProperties]
        if isinstance(_properties, Unset):
            properties = UNSET
        else:
            properties = TrinoParametersSpecProperties.from_dict(_properties)

        database = d.pop("database", UNSET)

        trino_parameters_spec = cls(
            trino_engine_type=trino_engine_type,
            host=host,
            port=port,
            user=user,
            password=password,
            aws_authentication_mode=aws_authentication_mode,
            athena_region=athena_region,
            catalog=catalog,
            athena_work_group=athena_work_group,
            athena_output_location=athena_output_location,
            properties=properties,
            database=database,
        )

        trino_parameters_spec.additional_properties = d
        return trino_parameters_spec

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties

File: dqops/src/main/java/com/dqops/connectors/trino/TrinoParametersSpec.java
/*
 * Copyright © 2023 DQOps (support@dqops.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dqops.connectors.trino;

import com.dqops.connectors.ConnectionProviderSpecificParameters;
import com.dqops.core.secrets.SecretValueLookupContext;
import com.dqops.core.secrets.SecretValueProvider;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMap;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMapImpl;
import com.dqops.metadata.sources.BaseProviderParametersSpec;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.EqualsAndHashCode;
import picocli.CommandLine;

import java.util.Collections;
import java.util.Map;
import java.util.Objects;

/**
 * Trino connection parameters.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@EqualsAndHashCode(callSuper = true)
public class TrinoParametersSpec extends BaseProviderParametersSpec
        implements ConnectionProviderSpecificParameters {
    private static final ChildHierarchyNodeFieldMapImpl<TrinoParametersSpec> FIELDS = new ChildHierarchyNodeFieldMapImpl<>(BaseProviderParametersSpec.FIELDS) {
        {
        }
    };

    @CommandLine.Option(names = {"--trino-engine"}, description = "Trino engine type.")
    @JsonPropertyDescription("Trino engine type. Supports also a ${TRINO_ENGINE} configuration with a custom environment variable.")
    private TrinoEngineType trinoEngineType;

    @CommandLine.Option(names = {"--trino-host"}, description = "Trino host name.")
    @JsonPropertyDescription("Trino host name. Supports also a ${TRINO_HOST} configuration with a custom environment variable.")
    private String host;

    @CommandLine.Option(names = {"--trino-port"}, description = "Trino port number.")
    @JsonPropertyDescription("Trino port number. The default port is 8080. Supports also a ${TRINO_PORT} configuration with a custom environment variable.")
    private String port;

    @CommandLine.Option(names = {"--trino-user"}, description = "Trino user name. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use dynamic substitution.")
    @JsonPropertyDescription("Trino user name. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use dynamic substitution.")
    private String user;

    // TODO: Handling authentication methods to different Trino instances: https://trino.io/docs/current/security/authentication-types.html
//    @CommandLine.Option(names = {"--trino-password"}, description = "Trino database password. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use dynamic substitution.")
//    @JsonPropertyDescription("Trino database password. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use dynamic substitution.")
//    private String password;

    @CommandLine.Option(names = {"-T"}, description = "Trino additional properties that are added to the JDBC connection string")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Map<String, String> properties;

    @CommandLine.Option(names = {"--athena-region"}, description = "The AWS Athena Region where queries will be run. Supports also a ${ATHENA_REGION} configuration with a custom environment variable.")
    @JsonPropertyDescription("The AWS Region where queries will be run. Supports also a ${ATHENA_REGION} configuration with a custom environment variable.")
    private String athenaRegion;

    @CommandLine.Option(names = {"--trino-catalog"}, description = "The Trino catalog that contains the databases and the tables that will be accessed with the driver. Supports also a ${TRINO_CATALOG} configuration with a custom environment variable.")
    @JsonPropertyDescription("The catalog that contains the databases and the tables that will be accessed with the driver. Supports also a ${TRINO_CATALOG} configuration with a custom environment variable.")
    private String catalog;

    @CommandLine.Option(names = {"--athena-work-group"}, description = "The Athena WorkGroup in which queries will run. Supports also a ${ATHENA_WORK_GROUP} configuration with a custom environment variable.")
    @JsonPropertyDescription("The workgroup in which queries will run. Supports also a ${ATHENA_WORK_GROUP} configuration with a custom environment variable.")
    private String athenaWorkGroup;

    @CommandLine.Option(names = {"--athena-output-location"}, description = "The location in Amazon S3 where query results will be stored. Supports also a ${ATHENA_OUTPUT_LOCATION} configuration with a custom environment variable.")
    @JsonPropertyDescription("The location in Amazon S3 where query results will be stored. Supports also a ${ATHENA_OUTPUT_LOCATION} configuration with a custom environment variable.")
    private String athenaOutputLocation;

    /**
     * Returns the trino engine type.
     * @return Trino engine type.
     */
    public TrinoEngineType getTrinoEngineType() {
        return trinoEngineType;
    }

    /**
     * Sets the trino engine type.
     * @param trinoEngineType New trino engine type.
     */
    public void setTrinoEngineType(TrinoEngineType trinoEngineType) {
        setDirtyIf(!Objects.equals(this.trinoEngineType, trinoEngineType));
        this.trinoEngineType = trinoEngineType;
    }

    /**
     * Returns the host name.
     * @return Host name.
     */
    public String getHost() {
        return host;
    }

    /**
     * Sets the host name.
     * @param host New host name.
     */
    public void setHost(String host) {
        setDirtyIf(!Objects.equals(this.host, host));
        this.host = host;
    }

    /**
     * Returns the port number. The value should store an environment variable expression or a numeric trino port number.
     * @return Port name or an expression to be extracted.
     */
    public String getPort() {
        return port;
    }

    /**
     * Sets the port number.
     * @param port Port name.
     */
    public void setPort(String port) {
        setDirtyIf(!Objects.equals(this.port, port));
        this.port = port;
    }

    /**
     * Returns the user that is used to log in to the data source (JDBC user or similar).
     * @return User name.
     */
    public String getUser() {
        return user;
    }

    /**
     * Sets a user name.
     * @param user User name.
     */
    public void setUser(String user) {
        setDirtyIf(!Objects.equals(this.user, user));
        this.user = user;
    }

//    /**
//     * Returns a password used to authenticate to the server.
//     * @return Password.
//     */
//    public String getPassword() {
//        return password;
//    }
//
//    /**
//     * Sets a password that is used to connect to the database.
//     * @param password Password.
//     */
//    public void setPassword(String password) {
//        setDirtyIf(!Objects.equals(this.password, password));
//        this.password = password;
//    }

    /**
     * Returns a key/value map of additional properties that are included in the JDBC connection string.
     * @return Key/value dictionary of additional JDBC properties.
     */
    public Map<String, String> getProperties() {
        return properties;
    }

    /**
     * Sets a dictionary of additional connection parameters that are added to the JDBC connection string.
     * @param properties Key/value dictionary with extra parameters.
     */
    public void setProperties(Map<String, String> properties) {
        setDirtyIf(!Objects.equals(this.properties, properties));
        this.properties = properties != null ? Collections.unmodifiableMap(properties) : null;
    }


    /**
     * Returns the Athena's AWS Region.
     * @return Athena's AWS Region.
     */
    public String getAthenaRegion() {
        return athenaRegion;
    }

    /**
     * Sets Athena's AWS Region.
     * @param athenaRegion Athena's AWS Region.
     */
    public void setAthenaRegion(String athenaRegion) {
        setDirtyIf(!Objects.equals(this.athenaRegion, athenaRegion));
        this.athenaRegion = athenaRegion;
    }

    /**
     * Returns the catalog
     * @return The catalog name.
     */
    public String getCatalog() {
        return catalog;
    }

    /**
     * Sets the catalog name.
     * @param catalog The catalog name.
     */
    public void setCatalog(String catalog) {
        setDirtyIf(!Objects.equals(this.catalog, catalog));
        this.catalog = catalog;
    }

    /**
     * Returns a high level container name that contian schemas.
     * @return High level container name that contian schemas.
     */
    @Override
    public String getDatabase(){
        return getCatalog();
    }

    /**
     * Returns the Athena's WorkGroup
     * @return Athena's WorkGroup.
     */
    public String getAthenaWorkGroup() {
        return athenaWorkGroup;
    }

    /**
     * Sets Athena's WorkGroup.
     * @param athenaWorkGroup Athena's WorkGroup.
     */
    public void setAthenaWorkGroup(String athenaWorkGroup) {
        setDirtyIf(!Objects.equals(this.athenaWorkGroup, athenaWorkGroup));
        this.athenaWorkGroup = athenaWorkGroup;
    }

    /**
     * Returns the Athena's OutputLocation
     * @return Athena's OutputLocation.
     */
    public String getAthenaOutputLocation() {
        return athenaOutputLocation;
    }

    /**
     * Sets Athena's OutputLocation.
     * @param athenaOutputLocation Athena's OutputLocation.
     */
    public void setAthenaOutputLocation(String athenaOutputLocation) {
        setDirtyIf(!Objects.equals(this.athenaOutputLocation, athenaOutputLocation));
        this.athenaOutputLocation = athenaOutputLocation;
    }

    /**
     * Returns the child map on the spec class with all fields.
     *
     * @return Return the field map.
     */
    @Override
    protected ChildHierarchyNodeFieldMap getChildMap() {
        return FIELDS;
    }

    /**
     * Creates and returns a deep copy of this object.
     */
    @Override
    public TrinoParametersSpec deepClone() {
        TrinoParametersSpec cloned = (TrinoParametersSpec)super.deepClone();
        return cloned;
    }

    /**
     * Creates a trimmed and expanded version of the object without unwanted properties, but with all variables like ${ENV_VAR} expanded.
     * @param secretValueProvider Secret value provider.
     * @param lookupContext Secret lookup context.
     * @return Trimmed and expanded version of this object.
     */
    public TrinoParametersSpec expandAndTrim(SecretValueProvider secretValueProvider, SecretValueLookupContext lookupContext) {
        TrinoParametersSpec cloned = this.deepClone();

        cloned.trinoEngineType = TrinoEngineType.valueOf(secretValueProvider.expandValue(cloned.trinoEngineType.toString(), lookupContext));

        cloned.host = secretValueProvider.expandValue(cloned.host, lookupContext);
        cloned.port = secretValueProvider.expandValue(cloned.port, lookupContext);
        cloned.user = secretValueProvider.expandValue(cloned.user, lookupContext);
        // cloned.password = secretValueProvider.expandValue(cloned.password, lookupContext);
        cloned.properties = secretValueProvider.expandProperties(cloned.properties, lookupContext);

        cloned.athenaRegion = secretValueProvider.expandValue(cloned.athenaRegion, lookupContext);
        cloned.catalog = secretValueProvider.expandValue(cloned.catalog, lookupContext);
        cloned.athenaWorkGroup = secretValueProvider.expandValue(cloned.athenaWorkGroup, lookupContext);
        cloned.athenaOutputLocation = secretValueProvider.expandValue(cloned.athenaOutputLocation, lookupContext);

        return cloned;
    }
}
